[WIP] :book: Add docs for writing provider integrations

janiskemper opened the pull request:
**What this PR does / why we need it**:
Adding docs for writing provider integrations

**Which issue(s) this PR fixes**:
Fixes #29

**TODOs**:

- [ ] squash commits
- [x] include documentation
- [ ] add unit tests

janiskemper commented on docs/PROVIDER.md:
```diff
@@ -0,0 +1,110 @@
+type ProviderClusterStackReleaseTemplateResource struct {
+	// +optional
+	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
+	Spec ProviderClusterStackReleaseSpec `json:"spec"`
```
I think you can have a look at CAPI's docs for provider implementations. They have the same structure. I cannot add the information you requested, because there is literally no pre-defined structure and we don't need to know anything about `ProviderClusterStackReleaseSpec`. 

I conclude for myself that this part is misleading more than helping, so we can think of removing this code example alltogether.

matofeder replied:
@janiskemper I agree that these code examples do not help the reader so much.

NotTheEvilOne replied:
I do not agree about the non-common structure issue. Please have a look at e.g. `runtime.RawExtension` which is exactly for that use case. You can give a specific k8s type with provider specific sections in place where providers can add their own structure needed.

[1] https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension
[2] https://github.com/kubernetes/apimachinery/issues/102

janiskemper replied:
so can you give me an example from the Cluster API context where this is used?

NotTheEvilOne replied:
See [1] this discussion for a good overview of the Cluster API decision. In short they replaced the need of provider specific fields by requiring the well known `<Provider><Type>` mechanism and referencing it.

Going one step back the first question to find out if `ProviderClusterStackRelease` is really expected to be named that way or if `<Provider>ClusterStackRelease` like `OpenStackClusterStackRelease` is meant instead. If the later is the case than no `RawExtension` use case exists but then it's the providers duty to report if a release is ready or not by patching the respective `status` field. That would be an excellent example of an common `cluster-stack-provider` Go package use case.

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20190709-cluster-spec-crds.md

janiskemper replied:
Maybe I completely misunderstand you, but I would say we name it exactly in this way, e.g. that for OpenStack it will be OpenStackClusterStackRelease.

janiskemper replied:
@matofeder after the conversation I had with multiple other people involved in the project of implementing a provider, I realized that this code example is actually really necessary. I have added a sentence now that `ProviderClusterStackReleaseSpec` has to be defined by the person implementing it to clarify it a bit better.

matofeder replied:
@janiskemper thx

jschoone replied:
Thanks for the additional sentence, @janiskemper, should make it more clear. I was wondering whether it makes sense to even write out the example, in this case
```
The prefix `Provider` should be replaced with the respective provider to build, e.g. for `Docker` it would be `DockerClusterStackReleaseSpec`.
```
These are details now, but would further limit the scope for interpretation.

Is there anything to add from you @NotTheEvilOne?
Otherwise we should mark this conversation as resolved

NotTheEvilOne replied:
The changes added here are sufficient, yes.

janiskemper replied:
and yes, @jschoone I can incorporate your changes!

janiskemper commented on docs/PROVIDER.md:
```diff
@@ -0,0 +1,110 @@
+
+In case that there are multiple node images which have to be provided to the user, it makes sense to have separate custom resources so that one custom resource is responsible for only one node image. A pattern where there is a `ProviderClusterStackRelease` that creates different `ProviderNodeImageRelease` custom resources, one for each node image that should be built, is advisable. 
+
+The same pattern that is also used in the objects of `ClusterStackRelease` and `ProviderClusterStackRelease` can be also applied here in a way that `ProviderClusterStackRelease` is set on `ready: true` if and only if all `ProviderNodeImageRelease` custom resources are also `ready: true`.
```
I don't agree here. Why should that be the case? Reliability of code will be mostly about how people write their code, not about how they name their API. And we cannot control how they write the code anyway. 
Maybe I misunderstand also what you mean, so you can elaborate maybe?

NotTheEvilOne replied:
It's about using Go's rudimentary support for interfaces to enforce / provide a certain structure for resources expected to be handled the same way for each provider implementation anyway not how the result is achieved using provider specific code that will be fundamentally different.

Take a look at CSI drivers as an example. They all implement common interfaces and work with storage classes. Commonly under parameters the filesystem type was and sometimes is defined as `fstype`. This meant that is was not always named the same way and moving from cloud to cloud got complicated. These days the `external-provisioner` handles PVC requests and storage classes. They provided the common name [1] `csi.storage.k8s.io/fstype` to handle the filesystem type the same way for all CSI drivers [2] and the common interface they share (it's used in e.g. `NodeStageVolume` under `req.VolumeCapability`).

[1] https://kubernetes-csi.github.io/docs/external-provisioner.html#storageclass-parameters
[2] https://github.com/kubernetes-csi/external-provisioner/blob/b255ee8f41384b330388c00a5f9252aff343fecf/pkg/controller/controller.go#L612

janiskemper replied:
We evaluated an interface approach, taking the Cloud Controller Manager as example. However, this is unfortunately not possible, same as with CAPI Provider Integrations, as provider-specific CRDs are needed!

matofeder commented on docs/PROVIDER.md:
```diff
@@ -0,0 +1,110 @@
+
+A Cluster Stack Provider should be built as a separate Kubernetes Operator.
+
+### Custom resources
```
IMO better to use some auto-generated docs for CRDs, e.g. just link this page https://doc.crds.dev/github.com/SovereignCloudStack/cluster-stack-operator

janiskemper replied:
the custom resources are not really "fixed", therefore we cannot add them to the reference unfortunately...

matofeder commented on docs/PROVIDER.md:
```diff
@@ -0,0 +1,110 @@
+## Implementing a new Provider Integration
+
+// TODO This has to be refined based on https://cluster-api.sigs.k8s.io/developer/providers/implementers-guide/overview.
+// The goal is that users can set up new repositories by themselves and know everything that is needed.
```
It seems that this part will be the most important, so I would recommend addressing it before we merge this PR.

matofeder reviewed (commented):
Overall, LGTM, just a few minor suggestions

janiskemper commented:
@michal-gubricky @NotTheEvilOne @jschoone I have added the documentation that we talked about in yesterday's meeting. I updated the previous section about best practices to make it clearer. @michal-gubricky if there are some open questions to you that I haven't answered yet, please tell me!

jschoone commented:
Closed because too old. Can be reopened if needed.
